distribution/lib/Standard/Base/0.0.0-dev/src/Data/Statistics.enso

```diff
@@ -388,3 +344,78 @@ is_valid v = case v of
+        Accumulator.Value self.count+1 minimum maximum (self.update_moments value)
+
+    ## PRIVATE
+    comparator current value bound = if current == Nothing then Nothing else
+        comp = Ordering.compare value current . catch Incomparable_Values (_-> Nothing)
+        if comp == Nothing then Nothing else
+            if comp == bound then value else current
+
+    ## PRIVATE
+    update_moments self value = if self.moments.is_nothing then Nothing else
+        case value of
+            _ : Number -> self.moments.map_with_index i->v-> v + (value ^ i+1)
+            _ -> Nothing
+
+    ## PRIVATE
+    compute : Statistic -> Any
+    compute self statistic =
+        case statistic of
+            Statistic.Count -> self.count
+            Statistic.Minimum -> if self.comparator_error then (Error.throw (Incomparable_Values.Error Nothing Nothing)) else self.minimum
+            Statistic.Maximum -> if self.comparator_error then (Error.throw (Incomparable_Values.Error Nothing Nothing)) else self.maximum
+            Statistic.Sum -> self.report_error statistic self.moments.first
+            Statistic.Mean -> self.report_error statistic (self.moments.first / self.count)
+            Statistic.Variance p -> self.report_error statistic (self.variance p)
+            Statistic.Standard_Deviation p -> (self.compute (Statistic.Variance p)) . sqrt
+            Statistic.Skew p -> self.report_error statistic (self.skew p)
+            Statistic.Kurtosis -> self.report_error statistic (self.kurtosis)
+
+    ## PRIVATE
+    variance self population =
+        min_points = if population then 1 else 2
+        if self.count < min_points then Number.nan else
+            denominator = if population then self.count else self.count-1
+            (self.moments.second - self.moments.first^2 / self.count) / denominator
+
+    ## PRIVATE
+    skew self population =
+        if self.count < 3 then Number.nan else
+            mean = self.compute (Statistic.Mean)
+            standard_deviation = self.compute (Statistic.Standard_Deviation population)
+            denominator = if population then self.count else ((self.count-1)*(self.count-2)/self.count)
+            scale = 1.0 / (standard_deviation^3) / denominator
+            ((self.moments.at 2) - 3 * mean * self.moments.second + 2 * mean * mean * self.moments.first) * scale
+
+    ## PRIVATE
+    kurtosis self =
+        if self.count < 4 then Number.nan else
+            n = self.count
+            mean = self.compute (Statistic.Mean)
+            variance = self.compute (Statistic.Variance False)
+            scale = n * (n+1) / ((n-1)*(n-2)*(n-3)*variance*variance)
+            shift = 3*(n-1)*(n-1)/((n-2)*(n-3))
+            ((self.moments.at 3) - 4 * mean * (self.moments.at 2) + 6 * mean * mean * self.moments.second - 3 * mean * mean * mean * self.moments.first) * scale - shift
+
+    ## PRIVATE
+    report_error self statistic ~action =
```

Comment:
very subjective and nitpicky but IMHO `check_error` could be a better name - `report` sounds like the error is always there